lib/index.js

```diff
@@ -0,0 +1,145 @@
+import React from 'react';
+import inherit from 'inherit';
+import renderTag from './renderTag';
+
+export default function bemReactCore(options, BaseComponent, classNameBuilder) {
+    const entities = {};
+    BaseComponent.prototype.__render = renderTag(classNameBuilder);
+
+    function applyEntityDecls() {
+        const entity = this;
+
+        if(entity.decls) {
+            entity.decls.forEach(({ fields, staticFields}) => {
+                entity.cls?
+                    inherit.self(entity.cls, fields, staticFields) :
+                    entity.cls = inherit(
+                        entity.base ? entity.base : BaseComponent,
+                        fields,
+                        {
+                            displayName : classNameBuilder.stringify(
+                                fields.block, fields.elem
+                            ),
+                            ...staticFields
+                        }
+                    );
+            });
+
+            entity.decls = null;
+        }
+
+        if(entity.modDecls) {
+            const ptp = entity.cls.prototype;
+
+            entity.modDecls.forEach(({ predicate, fields, staticFields }) => {
+                for(let name in fields) {
+                    const field = fields[name];
+                    typeof field === 'function' && (fields[name] = function() {
+                        let method;
+                        if(predicate.call(this, this.props)) {
+                            method = field;
+                        } else {
+                            const baseMethod = ptp[name];
+                            baseMethod && baseMethod !== field &&
+                                (method = this.__base);
+                        }
+
+                        return method && method.apply(this, arguments);
+                    });
+                }
+
+                inherit.self(entity.cls, fields, staticFields);
+            });
+
+            entity.modDecls = null;
+        }
+
+        return entity.cls;
+    }
+
+    function getEntity(key) {
+        return entities[key] || (entities[key] = {
+            cls : null,
+            base : null,
+            decls : null,
+            modDecls : null,
+            applyDecls : applyEntityDecls
+        });
+    }
+
+    return {
```

Comment:
maybe this file should be named `decls.js`? it contains only `decl*()` functions and this way we can avoid two `index.js` and `lib/index.js` with different meaning